workflow/sync/sync_manager.go

```diff
@@ -113,102 +111,146 @@ func (cm *Manager) Initialize(wfs []wfv1.Workflow) {
+		if !found {
+			var err error
+			switch syncItems[i].getType() {
+			case wfv1.SynchronizationTypeSemaphore:
+				lock, err = sm.initializeSemaphore(lockKey)
+			case wfv1.SynchronizationTypeMutex:
+				lock = sm.initializeMutex(lockKey)
+			default:
+				return false, false, "", failedLockName, fmt.Errorf("unknown Synchronization Type")
+			}
+			if err != nil {
+				return false, false, "", failedLockName, err
+			}
+			sm.syncLockMap[lockKey] = lock
+		}
 
-	if syncLockRef.GetType() == wfv1.SynchronizationTypeSemaphore {
-		err := cm.checkAndUpdateSemaphoreSize(lock)
-		if err != nil {
-			return false, false, "", err
+		if syncItems[i].getType() == wfv1.SynchronizationTypeSemaphore {
+			err := sm.checkAndUpdateSemaphoreSize(lock)
+			if err != nil {
+				return false, false, "", "", err
+			}
 		}
-	}
 
-	holderKey := getHolderKey(wf, nodeName)
-	var priority int32
-	if wf.Spec.Priority != nil {
-		priority = *wf.Spec.Priority
-	} else {
-		priority = 0
-	}
-	creationTime := wf.CreationTimestamp
-	lock.addToQueue(holderKey, priority, creationTime.Time)
-
-	ensureInit(wf, syncLockRef.GetType())
-	currentHolders := cm.getCurrentLockHolders(lockKey)
-	acquired, msg := lock.tryAcquire(holderKey)
-	if acquired {
-		updated := wf.Status.Synchronization.GetStatus(syncLockRef.GetType()).LockAcquired(holderKey, lockKey, currentHolders)
-		return true, updated, "", nil
+		var priority int32
+		if wf.Spec.Priority != nil {
+			priority = *wf.Spec.Priority
+		} else {
+			priority = 0
+		}
+		creationTime := wf.CreationTimestamp
+		lock.addToQueue(holderKey, priority, creationTime.Time)
+
+		ensureInit(wf, syncItems[i].getType())
+		acquired, already, newMsg := lock.checkAcquire(holderKey)
+		if !acquired && !already {
+			allAcquirable = false
+			if msg == "" {
+				msg = newMsg
+			}
+			if failedLockName == "" {
+				failedLockName = lockKey
+			}
+		}
 	}
 
-	updated := wf.Status.Synchronization.GetStatus(syncLockRef.GetType()).LockWaiting(holderKey, lockKey, currentHolders)
-	return false, updated, msg, nil
+	switch {
+	case allAcquirable:
+		updated := false
+		for i, lockKey := range lockKeys {
+			lock := sm.syncLockMap[lockKey]
+			acquired, msg := lock.tryAcquire(holderKey)
+			if !acquired {
+				return false, false, "", failedLockName, fmt.Errorf("bug: failed to acquire something that should have been checked: %s", msg)
+			}
+			currentHolders := sm.getCurrentLockHolders(lockKey)
+			if wf.Status.Synchronization.GetStatus(syncItems[i].getType()).LockAcquired(holderKey, lockKey, currentHolders) {
+				updated = true
+			}
+		}
+		return true, updated, msg, failedLockName, nil
+	default: // Not all acquirable
+		updated := false
+		for i, lockKey := range lockKeys {
+			currentHolders := sm.getCurrentLockHolders(lockKey)
+			if wf.Status.Synchronization.GetStatus(syncItems[i].getType()).LockWaiting(holderKey, lockKey, currentHolders) {
+				updated = true
+			}
+		}
+		return false, updated, msg, failedLockName, nil
+	}
```

Comment:
This would be more readable in a if statement with early return imo.

Reply:
I was under the impression `switch` was considered idiomatic go, so I've been trying to wean myself off ifs even for short ones.
Do you require it changing?

Reply:
No requirement, I will leave it up to you, because now we are effectively talking about subjective preferences.
I just think this pattern will read nicer:
```go
if cond {
  return true
}
return false
```